### PR TITLE
EN-4706 MW_RewardPoints - Fix the mismatch total amount issue if the customers redeem their maximum reward points

### PR DESCRIPTION
**Pull request by johnmccombs1**
# Description
This PR resolves the following case:

1. In the admin, go to Store -> Configuration -> Mageworld -> Reward Points Pro -> General.

Set Allow redeemed points to apply to tax to No.

Set Allow redeemed points to apply to shipping to Yes.

2. Add product to the cart.

3. On the cart page, choose to Maximize my discount with points, see here - https://prnt.sc/12j4c55

4. Open Bolt modal, estimate shipping and click on the Pay button. The error occurs because of the mismatch total amount issue. See the Bugsnag link here: https://app.bugsnag.com/bolt/magento-2/errors/6089c87aa91b6f000845c98b?filters%5Bevent.since%5D%5B%5D%5Btype%5D=eq&filters%5Bevent.since%5D%5B%5D%5Bvalue%5D=30d&filters%5Bsearch%5D%5B%5D%5Btype%5D=eq&filters%5Bsearch%5D%5B%5D%5Bvalue%5D=%22funandfunction.com%22&event_id=608e6fd60079c1259b7b0000 

See the video for replicating the issue here: https://drive.google.com/file/d/17gvnQDWEBLjPBhVilathKjVzQi3A5QHj/view

In this case, Magento just applies reward points for shipping, not for tax amount so the Magento total amount is the tax amount ($1.02). But Bolt applied reward points for both shipping and tax so the Bolt total amount is $0.

This PR resolves the issue by adjusting the shipping amount to correct the Bolt total amount..

Fixes: https://boltpay.atlassian.net/browse/EN-4706

#changelog EN-4706 MW_RewardPoints - Fix the mismatch total amount issue if the customers redeem their maximum reward points

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 
- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my Jira ticket link and provided a changelog message.

### CODE REVIEW

File: Model/EventsForThirdPartyModules.php [hunk 1]
```diff
@@ -758,8 +758,27 @@ class EventsForThirdPartyModules
                     "sendClasses" => ["Mirasvit\Rewards\Model\Config"],
                     "boltClass" => Mirasvit_Rewards::class,
                 ],
+                [
```
Comment (danac-gs) [inline Model/EventsForThirdPartyModules.php:761]:
```suggestion
                'MW_RewardPoints' => [
```
for easier navigation

File: Model/EventsForThirdPartyModules.php [hunk 2]
```diff
@@ -758,8 +758,27 @@ class EventsForThirdPartyModules
...
+                    "sendClasses" => [
+                        "MW\RewardPoints\Helper\Data",
+                    ],
+                    'boltClass'   => MW_RewardPoints::class,
+                ],
             ]
         ],
+        "filterShippingAmount" => [
+            'listeners' => [
+                [
```
Comment (danac-gs) [inline Model/EventsForThirdPartyModules.php:772]:
```suggestion
                'MW_RewardPoints' => [
```

File: ThirdPartyModules/MageWorld/RewardPoints.php [hunk 3]
```diff
@@ -246,4 +246,76 @@ public function removeAppliedStoreCredit(
...
+     * @param $mwRewardPointsHelperData
+     * @param $mwRewardPointsModelCustomer
+     * @param $quote
+     * @return int
+     */
+    public function filterShippingAmount($result,
+                                         $mwRewardPointsHelperData,
+                                         $mwRewardPointsModelCustomer,
+                                         $quote)
+    {
```
Comment (danac-gs) [inline ThirdPartyModules/MageWorld/RewardPoints.php:261]:
```suggestion
    /**
     * @param int                                    $result
     * @param \MW\RewardPoints\Helper\Data           $mwRewardPointsHelperData
     * @param \MW\RewardPoints\Model\CustomerFactory $mwRewardPointsModelCustomerFactory
     * @param \Magento\Quote\Model\Quote             $quote
     *
     * @return int
     */
    public function filterShippingAmount(
        $result,
        $mwRewardPointsHelperData,
        $mwRewardPointsModelCustomerFactory,
        $quote
    ) {
```

File: ThirdPartyModules/MageWorld/RewardPoints.php [hunk 4]
```diff
@@ -246,4 +246,76 @@ public function removeAppliedStoreCredit(
...
+        $this->mwRewardPointsHelperData = $mwRewardPointsHelperData;
+        $this->mwRewardPointsModelCustomer = $mwRewardPointsModelCustomer;
```
Comment (danac-gs) [inline ThirdPartyModules/MageWorld/RewardPoints.php:263]:
```suggestion
        $this->mwRewardPointsModelCustomer = $mwRewardPointsModelCustomerFactory;
```

File: ThirdPartyModules/MageWorld/RewardPoints.php [hunk 5]
```diff
@@ -246,4 +246,76 @@ public function removeAppliedStoreCredit(
...
+     * @param $quote
+     * @param $transaction
+     * @return bool
+     */
+    public function filterSkipValidateShippingForProcessNewOrder(
+        $result,
+        $mwRewardPointsHelperData,
+        $quote,
+        $transaction
+    ) {
```
Comment (danac-gs) [inline ThirdPartyModules/MageWorld/RewardPoints.php:302]:
```suggestion
    /**
     * @param bool                         $result
     * @param \MW\RewardPoints\Helper\Data $mwRewardPointsHelperData
     * @param \Magento\Quote\Model\Quote   $quote
     * @param \stdClass                    $transaction
     *
     * @return bool
     */
    public function filterSkipValidateShippingForProcessNewOrder(
        $result,
        $mwRewardPointsHelperData,
        $quote,
        $transaction
    ) {
```

**Review by danac-gs — APPROVED** [review]
LGTM. Added some cosmetic changes.

**Review by njfaries — APPROVED** [review]
LGTM. Please resolve outstanding comments.

**Comment by johnmccombs1** [comment]
@njfaries  @danac-gs I've updated the code. Can you re-check?

**Pull request merged**